development/tools/cmd/oomfinder/main.go

```diff
@@ -5,118 +5,154 @@ import (
 	"time"
 
 	"github.com/containerd/containerd"
+	events2 "github.com/containerd/containerd/api/events"
+	"github.com/containerd/containerd/events"
 	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/typeurl"
 	dockerclient "github.com/fsouza/go-dockerclient"
 	log "github.com/sirupsen/logrus"
+	flag "github.com/spf13/pflag"
+
+	// Register grpc event types
+	_ "github.com/containerd/containerd/api/events"
 )
 
-// dockerOOMListener create channel to get events from docker daemon.
+var debug = flag.BoolP("debug", "d", false, "enable debug output")
+
+// dockerOOMListener listen for events from docker daemon.
 // Listening is done in a goroutine.
 // Channel will receive only oom events as defined in Filters property of EventsOptions.
-// On oom event details are printed to stdout.
+// On oom event, details are printed to stdout.
 func dockerOOMListener(client *dockerclient.Client, wg *sync.WaitGroup) {
 	eventsChannel := make(chan *dockerclient.APIEvents)
+	// Filter events to get only OOM events for containers.
 	eventOptions := dockerclient.EventsOptions{
 		Filters: map[string][]string{
 			"type":  {"container"},
 			"event": {"oom"},
 		},
 	}
-	errs := client.AddEventListenerWithOptions(eventOptions, eventsChannel)
-	if errs != nil {
-		log.WithFields(log.Fields{"msg": "got docker daemon events error"}).Errorf("%s", errs)
+	// Subscribe to listen for docker deamon events.
+	err := client.AddEventListenerWithOptions(eventOptions, eventsChannel)
+	if err != nil {
+		log.WithError(err).Error("got docker daemon events error")
 	}
+	log.Debug("Subscribed on docker socket")
 	go func() {
 		defer wg.Done()
 		for {
 			event, ok := <-eventsChannel
 			if !ok {
 				log.Error("failed read docker event")
+				return
 			}
-			fmt.Fprintf(os.Stdout, "OOM event received time: %s , namespace: %s , pod: %s ,container: %s , image: %s \n",
+			log.Debugf("Got docker oom event: %v", event)
+			_, err := fmt.Fprintf(os.Stdout, "OOM event received, time: %s, namespace: %s, pod: %s, container: %s, image: %s\n",
 				time.Unix(event.Time, 0).Format(time.RFC822Z),
 				event.Actor.Attributes["io.kubernetes.pod.namespace"],
 				event.Actor.Attributes["io.kubernetes.pod.name"],
 				event.Actor.Attributes["io.kubernetes.container.name"],
 				event.Actor.Attributes["image"],
 			)
+			if err != nil {
+				log.WithError(err).Error("cannot print event details")
+				continue
+			}
 		}
 	}()
 }
 
-// containerOOMListener create channel to get events from containerd daemon.
-// Listening is done within to goroutines. One for errors and second one for events.
+// containerOOMListener listen for events from containerd daemon.
+// Listening is done within goroutine.
 // Channels will receive only oom events as defined in filters argument of Subscribe method.
-// On oom event details are printed to stdout.
+// On oom event details, are printed to stdout.
 func containerdOOMListener(client *containerd.Client, wg *sync.WaitGroup) {
+	defer wg.Done()
```

Comment:
Having the `defer` here will immediately execute once you hit the `go func()`. This will mean the thread will be running, but will not be registered. Either run the entire function as stated before, or move this part inside the grouting spawn.

Reply:
You right. Don't know why I moved it here. Thanks.